data/posts/remix-i18n.md

```diff
@@ -0,0 +1,504 @@
+
+- Catch all slugs in nested levels
+- By just adding `($lang)` in the routes enables to catch `lang` parameter in the route
+
+Also, it's possible to detect if `params.lang` is not a valid language value by creating a reusable helper function. It's a great way to provide the best UX for users.
+
+Let's look more in-depth at what I mentioned with the example code. You can fork and clone this [Optional Segments example app repo][optional-segments-example-repo] to test it out on your local machine.
+
+![A GIF of Remix Optional Segments example app][optional-segments-app-gif]
+
+This example app is partially based on the Remix contact app tutorial. It has a `contacts` route and `contacts` route has a `contactId` parameter.
+
+```
+ app/
+├── components/
+│   └── Header.tsx // language switcher
+├── routes/
+│   ├── _index.tsx
+│   ├── ($lang).contacts.$contactId.tsx
+│   └── ($lang).contacts.tsx
+├── root.tsx
+├── data.tsx // contacts data
+└── utils.ts // reusable getLang function to check valid params.lang
+```
+
+A `lang` parameter is where you can see it catches all slugs in different nested levels, such as `ja/contacts` and `ja/contacts/ryan-florance` in this example app. It covers the case if you want to implement localized sub-directories without CMSs.
+
+Built-in parameters like `params.lang` saves you time when implementing i18n-supported routes. To enable Optional Segments, you can add `($lang)` in the routes like this, `app/routes/($lang).contacts.$contactId` to catch `lang` parameter in the route.
+
+Configuring localized sub-directories with functionally working routes is the most important yet time-consufnig part of the i18n implementation process depending on the built-in features from frameworks.
+Remix removes the pain by providing helpful parameters, a flexible structure, and the best DX for developers.
+
+> Make sure you follow the [Google SEO guideline][google-seo-guideline] for non-ASCII characters in the URL. It's not recommended to use non-ASCII characters in the URL. (i.e., `ja/contacts/マイケル-ジャクソン`) It's better to use ASCII characters in the URL. (i.e., `ja/contacts/michael-jackson`)
+
+The Optional Segments example app repo also includes a reusable function to check if `params.lang` is not a valid language code. For better UX, detecting when users access the URL with an invalid language code is essential. It is also important to show them that the page with an invalid language slug doesn't exist.
+
+```js
+// utils.ts
+import { Params } from "@remix-run/react";
+
+export function getLang(params: Params<string>) {
+  const lang = params.lang ?? "en";
+  if (lang !== "ja" && lang !== "en") {
+    throw new Error(`Invalid language ${lang}`);
+  }
+  return lang;
+}
+```
+
+As this `getLang` function is also designed to return selected valid `params.lang` values, it can be used to get only the necessary data for the selected language. The contact data object with types in `data.tsx` looks like this.
```

Comment:
```suggestion
The `getLang` function returns the selected valid `params.lang` value, which means it can be used to get only the necessary data for the selected language. The contact data object with types in `data.tsx` looks like this:
```